### PR TITLE
Adding argument to GenotypeGVCFs to keep only RAW_GT_COUNT

### PR DESCRIPTION
**Pull request by meganshand**
Currently, there is an argument to keep all raw annotations, but the flow based pipeline has a use case that needs to keep `RAW_GT_COUNT` without keeping the other raw annotations. This new argument allows the user to keep `RAW_GT_COUNT` while still cleaning up the other raw annotations in GenotypeGVCFs.

### CODE REVIEW

**Comment by samuelklee** [comment]
As discussed, let's quickly see if we can generalize this to allow one to specify a set of raw annotations to keep. If it looks like that will be a reach, just let me know and I can go ahead and review. Thanks!

**Comment by meganshand** [comment]
@samuelklee I made an attempt to make the argument more general. It should be ready for review. Thanks!

**Comment by samuelklee** [comment]
Thanks, @meganshand! Just a heads up, I’m mostly out this week since it’s a vacation week for our daycare. But will try to sneak this in if I can.

**Review by samuelklee — APPROVED** [review]
Thanks @meganshand! Sorry for the delay. Some minor changes, and some potentially major/breaking ones that might merit more discussion.

If you need this merged quickly, I'll trust you to make some of the corresponding decisions. But feel free to ping me if you'd like to discuss!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java
Comment (meganshand) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java]: This is finally resolved! I got help from @cmnbroad to add an argument that uses the `GATKAnnotationPlugin`, but even so I think I did not implement this in the cleanest way. @cmnbroad I'd appreciate if you'd take a look at this branch again to make sure I didn't miss anything.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:933]: Sorry, just to be clear, you mean you'll revert this test to its original behavior, right? In that case, will you add another test for the keep-specified behavior?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java
Comment (meganshand) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:940]: The issue is that `RawGtCount` is combined by GenomicsDB as a sum. There is no `RawGtCount` annotation on RefBlocks so the HomRef count always ends up being 0 in this case. It seems like `ExcessHet` isn't actually using `RawGtCount` as I expected. It recalculates the counts in `GenotypeUtils.computeDiploidGenotypeCounts` which doesn't seem to use `RawGtCount` so I'm not sure why `RawGtCount` is the `rawKey` associated with `ExcessHet`. Seems like this whole thing could be cleaned up so that we use the same method `ExcessHet` does to calculate the count of Het and HomVar genotypes at a site, but I'm not sure why we didn't do that in the first place.

Also I added the documentation to the header line.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java
Comment (meganshand) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:933]: The original behavior of `--keep-combined-raw-annotations` set to true while `--keep-specified-raw-annotations` is not empty is no longer allowed due to the `mutex`. So I think I am not reverting this test to the original behavior. Does that make sense? This test sets `--keep-specified-raw-annotations` to `RawGtCount`

**Comment by meganshand** [comment]
I think tests are failing only because I changed a header line. I will update the expected VCFs.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:933]: Sorry, by "original behavior" I mean the current behavior of the test as it is in master. This test sets `--keep-combined-raw-annotations true` and hence additionally retains `RAW_MQandDP`.

In contrast, you modify the test in your branch so that only `RAW_GT_COUNT` is retained. This means the code path that is responsible for retaining `RAW_MQandDP` is no longer run in this test, at least (and is perhaps not run for this particular input test file in any test at all). Even though the original test does not explicitly check for any particular `RAW_MQandDP` output, this is still an effective drop in test coverage.

Probably not anything to sweat about, but that's all that I wanted to point out in my original comment. I'll leave it up to you to decide if `--keep-combined-raw-annotations true` is sufficiently covered by the other tests so that we don't have to worry about replacing the original test here.

Comment (meganshand) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:933]: Oh! Sorry, I see what you're saying now. I think I originally added this test as a way of testing `RawGTCount`, rather than `--keep-combined-raw-annotations true`, so I overlooked that it did actually test the latter too. I do think the other tests sufficiently cover `--keep-combined-raw-annotations true` but I very much see how that is not clear from this change.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java:281]: I think it would be preferable to add an override of `makeVariantAnnotations` into this class (adjacent to the `getPluginDescriptors` override), and move the computation of which annotations to keep into that, rather than inlining it here.

Also, you're basically giving priority to the new arg in the case where it conflicts with, say, an exclude arg, which I think is fine if thats what you want, but you might want to document that as part of the new arg if thats what you go with.

File: src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKAnnotationArgumentCollection.java
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKAnnotationArgumentCollection.java:42]: Also this can be removed.

File: src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKAnnotationPluginDescriptor.java
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKAnnotationPluginDescriptor.java:499]: If we keep this method, I'd suggest:

- moving it so its adjacent to `getResolvedInstances`
- renaming it to `getDiscoveredInstances`
- adding a javadoc comment saying that it returns *every* annotation that is discovered anywhere on the classpath, and that it should only be used when `getResolvedInstances` isn't sufficient

**Review by cmnbroad — CHANGES_REQUESTED** [review]
@meganshand Some changes requested, hope they make sense. I didn't really look at the VariantAnnotatorEngine changes, but presumably someone else is reviewing those.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java:232]: Also, make sure that when you add the instance variable, you don't annotate it as an `@ArgumentCollection` (and maybe even add a comment saying that `@ArgumentCollection` is left out on purpose) since if you do the parser will complain about duplicate arg names.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java
Comment (meganshand) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: The problem is I still need the regular list of annotations (in addition to the `annotationsToKeep`). I can still move it to a separate function though. 

Also I think I resolved this by not using `allDiscoveredAnnotations` so now excluded args should get handled correctly.

**Review by meganshand — COMMENTED** [review]
Thank you @cmnbroad! I knew it wasn't right, but I was missing the key step of declaring the Argument Collection in GenotpeGVCFs without labeling it an ArgumentCollection. 

Hopefully it's cleaner now and should be ready for re-review.

**Review by cmnbroad — CHANGES_REQUESTED** [review]
@meganshand The argument stuff looks much better now, since everything specific to GenotypeGVCFs is private to it now, but see my comments about the list reconciliation.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java
Comment (meganshand) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: I moved the "missing annotation" check here and added a test.

**Comment by meganshand** [comment]
Thanks @cmnbroad, back to you!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java:79]: One other minor nit: why introduce this intermediate alias ? To me having the additional alias with a slightly different name just obfuscates whats going on - I'd suggest just allocating and using the `rawAnnotationsToKeep` instance variable directly, since this winds up being assigned there anyway.

**Review by cmnbroad — COMMENTED** [review]
@meganshand I mostly focused on the arg part (which looks good now - phew!) on the theory that Sam reviewed the rest of this. I added one more nit/comment, up to you if you want to use it or not. Otherwise LGTM.

**Comment by meganshand** [comment]
@cmnbroad I think Sam has reviewed the rest of this PR. Thank you for helping me get the argument part working! Assuming tests pass this is ready for a final review.

**Comment by samuelklee** [comment]
@meganshand you've got a prior approval from me, but let me know if  you think I should re-review. Thanks @cmnbroad for helping out with the argument code and reviewing!

**Review by cmnbroad — APPROVED** [review]
LGTM.

**Pull request merged**
